### ISSUE TITLE
Expose functions to create mesh array from any `Mesh` primitive

### ISSUE THREAD

**Issue by Scony** [godotengine/godot-proposals#3918]
### Describe the project you are working on

Strategy game with elements of RPG. Heavily reliant on navigation.

### Describe the problem or limitation you are having in your project

Cannot bake static navigation obstacles based on static colliders in reasonable time due to this bug: https://github.com/godotengine/godot/issues/57148

TL;DR - At the moment, when baking on a separate thread and gathering the geometry for static colliders, temporary primitive mesh instances need to be created so that `ArrayMesh` data can be obtained. Unfortunately, in the case of a separate thread, an enormous amount of semaphore locking is required as all the calls are going through `VisualServer` (i.e. uploading triangles, operating on triangles, fetching triangles etc.).

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I propose exposing (at least internally) functions for creating mesh array data (raw triangles basically) for each primitive mesh so that one can obtain triangles without any calls going through `VisualServer`.

As a side effect, the low-level mesh computation algorithm could be re-used in different engine places like e.g. `CSG` implementation.

As another side effect, it would be super simple to expose those functions further - to the end-user - via some kind of API similar to `MeshDataTool` / `SurfaceTool` so that it would enable composing complex meshes from primitive shapes - not just bare triangles/vertices.

This enhancement would also help with #2249

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

For each primitive mesh there would be a new `public` `static` function called `create_mesh_array` which would calculate and fill provided `Array` with mesh data. Alternatively, it could create such an `Array` and return it.

So e.g. for `CubeMesh` we would have:
```
static void create_mesh_array(Array &p_arr, Vector3 size, int subdivide_w, int subdivide_h, int subdivide_d);
```

That in turn would simplify e.g. collecting navmesh geometry as drafted here (for `CubeMesh` only for now): https://github.com/godotengine/godot/pull/57411/files

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No.

### Is there a reason why this should be core and not an add-on in the asset library?

It's required for threaded navmesh baking which is part of core.

**Comment by BastiaanOlij** [comment]
I discussed this issue with @Scony and recommended he raise this proposal to get further community feedback.

The needs behind this proposal are sound. The logic for creating the mesh data for primitive shapes is currently tucked away in the primitive mesh classes. We have already got some duplication of logic with the CSG nodes which create similar shapes with similar code and there are other use cases where you would like to obtain the raw data for use elsewhere. In that I endorse this proposal as the original author of the primitive mesh classes.

The discussion that I believe we should have community input for is how we would like to see this implemented. The simplest solution is to make the current logic accessible through static methods as suggested in the OP but it may not be the neatest solution.

My current gut feeling is that the proper solution involves creating a tool class that produces the raw data for the various mesh types and have the primitives mesh classes be consumers of this tool class. This may also involve evaluating if the logic is better implemented in closer harmony to the surface tool or if the current approach remains the optimum (which is based on the format of data the VisualServer expects but which is more an end product format).

**Comment by fire** [comment]
I have a few questions.

Is this something like a ImmediateMesh?

https://docs.godotengine.org/en/latest/classes/class_immediatemesh.html#class-immediatemesh

**Comment by Scony** [comment]
@fire the exposed logic would enable very simple implementation of such a imaginary function for `SurfaceTool`:

> void append_cube( Vector3 size, int subdivide_w, int subdivide_h, int subdivide_d, int surface, Transform3D transform )
> Create a cube and append its vertices to the current vertex array with specified `Transform3D`

**Comment by fire** [comment]
Is this code not sufficient?

```gdscript
var box : BoxMesh = BoxMesh.new()
var st : SurfaceTool = SurfaceTool.new()
st.begin(Mesh.PRIMITIVE_TRIANGLES)
st.append_from(box, 0,)
var mesh : ArrayMesh = st.commit()
```

**Comment by fire** [comment]
I think I get what you mean, but I don't think it's a good to duplicate the current primitive meshes.

Unless it's possible to have these options as a ArrayMesh enum and settings which may work.

1. Remove PrimitiveMesh
2. Go to ArrayMesh
3. For each type of PrimitiveMesh add a method for `append_cube( Vector3 size, int subdivide_w, int subdivide_h, int subdivide_d, int surface, Transform3D transform )` etc.
4. For the editor inspector select a primitive enum and provide options.

**Comment by fire** [comment]
Alternative design.

https://github.com/godotengine/godot-proposals/issues/3918#issuecomment-1030801992

Let's ignore touching the PrimitiveMeshes.

1. Move all the code for like `append_cube` to SurfaceTool
2. BoxMesh can internally call SurfaceTool
3. CSG etc can also call SurfaceTool

**Comment by Zireael07** [comment]
Why do we even have PrimitiveMesh that's distinct from ArrayMesh 🤔

**Comment by Calinou** [comment]
> Why do we even have PrimitiveMesh that's distinct from ArrayMesh thinking

See https://github.com/godotengine/godot-proposals/issues/2249#issuecomment-826519987.

**Comment by Scony** [comment]
> Is this code not sufficient?
> 
> ```gdscript
> # Untested code
> var box : BoxMesh = BoxMesh.new()
> var st : SurfaceTool = SurfaceTool.new()
> st.begin(Mesh.PRIMITIVE_TRIANGLES)
> st.append_from(box, 0)
> var mesh : ArrayMesh = st.commit()
> ```

I encourage you to try adding 1000 such boxes in a loop while being on a separate thread. Curious how long it will take.

In general, the code above is more/less how currently the geometry is being collected for the purpose of navigation mesh calculation. Unfortunately, the above code needs to lock `VisualServer` semaphore >3 times. And since that semaphore is locked by the main thread very often (due to the main game loop) it will be a very hard bottleneck. You'll basically run into this: https://github.com/godotengine/godot/issues/56524

This is why we don't want to involve actual Meshes (which are tied to `NavigationServer`).

**Comment by fire** [comment]
Is this still a limitation in Godot 4 or we designing this api for Godot 3?

**Comment by Scony** [comment]
@fire 

Actually, I haven't checked on `master` - tested on `3.x`, but I assume it's the same.

You can take a look at this comment: https://github.com/godotengine/godot/issues/57148#issuecomment-1021626938 to see why we're running into locks exactly.

**Comment by fire** [comment]
Since this appears to be an optimization and you can do this already in Godot using regular apis. Can you provide some benchmarks on Godot Engine 4 that show it's too slow for the navigation server usecase?

**Comment by Zireael07** [comment]
@fire there's at least two issues open where things are extremely slow IF they are on a thread and I guess the OP was wanting to do navmesh stuff on a thread too (as it seems like an obvious thing to do if you want to optimize game start/setup)

**Comment by Scony** [comment]
@fire sure, I'll provide some when the time comes. For now let's focus on `3.x` as this will be required for `3.5` release.

**Comment by BastiaanOlij** [comment]
> Why do we even have PrimitiveMesh that's distinct from ArrayMesh 🤔
Two very different use cases. ArrayMesh is an all purpose class where the user can load any mesh data into. The PrimitiveMesh classes were written to quickly add standard meshes into your scene. In the end they use the same data structure that gets sent to the visual server but the source of the data is very different.

Also remember that I created the PrimitiveMesh classes as one of the first things I added to Godot so a lot of use cases at the time hadn't been discovered yet. They were added because there was a large demand from the community to have more than the standard cube that was available at the time mostly for the purpose of rapid prototyping a level. They predate CSG and it's a bit of a shame the need for this change wasn't identified at the time and instead the CSG classes duplicated a lot of the logic that was in the primitive meshes.  

> Since this appears to be an optimization and you can do this already in Godot using regular apis. Can you provide some benchmarks on Godot Engine 4 that show it's too slow for the navigation server usecase?

Keep in mind that performance is not the only concern here. The primitive meshes send all their data to the visual server which results in buffers being allocated in OpenGL/Vulkan, data being transferred to the GPUs memory, etc.
All for nothing because we're not interested in ever using the primitive meshes, we're just interested in the source data to use somewhere else in the system. 

So it is a lot of overhead in a lot of places.
